src/python/pants/backend/python/providers/pyenv/rules.py

```diff
@@ -0,0 +1,183 @@
+                        if [ ! -d "$DEST" ]; then
+                            export PYENV_ROOT=$(realpath $(mktemp -d -u -p {PYENV_NAMED_CACHE} {specific_python}.XXXXXX))
+                            # export LDFLAGS="$LDFLAGS -Wl,-rpath=PYENV_ROOT/lib"
+                            {pyenv.exe} install {specific_python}
+                            # Removing write perms helps ensure users aren't accidentally modifying Python
+                            # or the site-packages
+                            chmod -R -w "$PYENV_ROOT"/versions/{specific_python}
+                            chmod +w "$PYENV_ROOT"/versions/{specific_python}
+                            ln -s "$PYENV_ROOT"/versions/{specific_python} {PYENV_NAMED_CACHE}/{specific_python}
+                            rm -rf "$PYENV_ROOT"/shims
+                        fi
+                        echo "$DEST"/bin/python
+                        """
+                    ).encode("ascii"),
+                    is_executable=True,
+                )
+            ]
+        ),
+    )
+    digest = await Get(Digest, MergeDigests([shim_digest, pyenv.digest]))
+
+    # NB: We don't cache this process at any level for two reasons:
+    #   1. Several tools (including pex) refer to Python at an absolute path, so a named cache is
+    #   the only way for this to work reasonably well. Since the named cache could be wiped between
+    #   runs (technically during a run, but we can't do anything about that) the
+    #   fastest-yet-still-correct solution is to always run this process and make it bail
+    #   early-and-quickly if the requisite Python already exists.
+    #   2. Pyenv compiles Python using whatever compiler the system is configured to use. Python
+    #   then stores this information so that it cn use the same compiler when compiling extension
+    #   modules. Therefore caching the compiled Python is somewhat unsafe (especially for a remote
+    #   cache).
+    result = await Get(
+        ProcessResult,
+        Process(
+            ["./install_python_shim.sh"],
+            input_digest=digest,
+            description=f"Install Python {python_to_use}",
+            append_only_caches=PYENV_APPEND_ONLY_CACHES,
+            env={
+                "PATH": env_vars.get("PATH", ""),
+                "TMPDIR": "{chroot}/tmpdir",
+                "LDFLAGS": env_vars.get("LDFLAGS", ""),
+                "PYTHON_CONFIGURE_OPTS": " ".join(pyenv_subsystem.python_configure_opts),
+            },
+            # Don't cache, we want this to always be run so that we can assume for the rest of the
+            # session the named_cache destination for this Python is valid, as the Python ecosystem
+            # mainly assumes absolute paths for Python interpreters.
+            cache_scope=ProcessCacheScope.PER_SESSION,
+        ),
+    )
+
+    return PythonExecutable(
+        path=result.stdout.decode("utf-8").splitlines()[-1].strip(),
```

Comment:
Just `decode()` will use UTF-8, and that seems to be how we do this idiomatically elsewhere in the codebase.

In fact, you could use `encode()` and `decode()` above and drop the `"ascii"`, since UTF-8 is a superset of ascii and you don't expect any non ascii characters...

Reply:
But again, not worth triggering a new CI run for.